text/0031-configuration.md

```diff
@@ -0,0 +1,287 @@
+- provide a `finish` method constructing the actual configuration object;
+- have default values defined as `const` and set with `Option::unwrap_or`/`Option::unwrap_or_else`;
+
+Here is a small example fitting all these requirements.
+
+`config.toml`
+```toml
+[snapshot]
+meta_file_path  = "./data/snapshot/mainnet.snapshot.meta"
+state_file_path = "./data/snapshot/mainnet.snapshot.state"
+```
+
+`config.rs`
+```rust
+const DEFAULT_META_FILE_PATH: &str = "./data/snapshot/mainnet.snapshot.meta";
+const DEFAULT_STATE_FILE_PATH: &str = "./data/snapshot/mainnet.snapshot.state";
+
+#[derive(Default, Deserialize)]
+pub struct SnapshotConfigBuilder {
+    meta_file_path: Option<String>,
+    state_file_path: Option<String>,
+}
+
+impl SnapshotConfigBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn meta_file_path(mut self, meta_file_path: String) -> Self {
+        self.meta_file_path.replace(meta_file_path);
+        self
+    }
+
+    pub fn state_file_path(mut self, state_file_path: String) -> Self {
+        self.state_file_path.replace(state_file_path);
+        self
+    }
+
+    pub fn finish(self) -> SnapshotConfig {
+        SnapshotConfig {
+            meta_file_path: self.meta_file_path.unwrap_or(DEFAULT_META_FILE_PATH.to_string()),
+            state_file_path: self.state_file_path.unwrap_or(DEFAULT_STATE_FILE_PATH.to_string()),
+        }
+    }
+}
+```
+
+### Configuration
+
+A configuration type should:
+- have a name suffixed by `Config`;
+- derive the following traits;
+  - `Clone` to be able to provide ownership to different components;
+  - `Serialize` if the configuration is expected to be updated and saved;
+- provide a `build` method that returns a new instance of the associated builder;
+- have the same fields with the same names, without `Option`, as the builder;
+- have no public fields;
+- provide setters/updaters only on fields that are expected to be updatable;
+- have getters or `pub(crate)` fields;
+
+Here is a small example fitting all these requirements:
+
+`config.toml`
+```toml
+[snapshot]
+meta_file_path  = "./data/snapshot/mainnet.snapshot.meta"
+state_file_path = "./data/snapshot/mainnet.snapshot.state"
+```
+
+`config.rs`
+```rust
+#[derive(Clone)]
+pub struct SnapshotConfig {
+    meta_file_path: String,
+    state_file_path: String,
+}
+
+impl SnapshotConfig {
+    pub fn build() -> SnapshotConfigBuilder {
+      SnapshotConfig::new()
+    }
+
+    pub fn meta_file_path(&self) -> &String {
+        &self.meta_file_path
+    }
+
+    pub fn state_file_path(&self) -> &String {
+        &self.state_file_path
+    }
+}
+
+```
+
+### Read a configuration builder from a file
+
+```rust
+let config_builder = match fs::read_to_string("config.toml") {
+    Ok(toml) => match toml::from_str::<SnapshotConfigBuilder>(&toml) {
+        Ok(config_builder) => config_builder,
+        Err(e) => {
+            panic!("Error parsing config file: {:?}", e);
```

Comment:
At some point we don't want to panic here (thinking of hot-loading configs when the user makes changes to it just as in Hornet, and if he doesn't edit the file correctly, the node shouldn't implode, but instead log an error, and reject updating its parameters). So maybe you could mention that in the "Unresolved questions" section?

Reply:
This is really just to show how to use the `toml` lib, this is not intended  to be copied as is.

Reply:
I can probably add a note.

Reply:
Additionally when the nodes shut down, it should write its internal parameters into the config. If it hasn't access to the file, because the user has the file open, the log should show an error that the node can't properly shutdown until it has access to the config.toml. That is to prevent to ever have an unparseable config.

Reply:
This is very, very implementation detail IMO. This RFC is just about configuration, it shouldn't speak about the node at all.

Reply:
I agree. I think I would still remove the explicit code in the error handle, and just put a comment there like:
```rust
// deal with parsing error
```
what do you think?

Reply:
Sounds good to me 👍